code/modules/actionspeed/_actionspeed_modifier.dm

```diff
@@ -0,0 +1,177 @@
+	5. Update if necessary
+*/
+/mob/proc/add_or_update_variable_actionspeed_modifier(datum/actionspeed_modifier/type_id_datum, update = TRUE, multiplicative_slowdown)
+	var/modified = FALSE
+	var/inject = FALSE
+	var/datum/actionspeed_modifier/final
+	if(istext(type_id_datum))
+		final = LAZYACCESS(actionspeed_modification, type_id_datum)
+		if(!final)
+			CRASH("Couldn't find existing modification when provided a text ID.")
+	else if(ispath(type_id_datum))
+		if(!initial(type_id_datum.variable))
+			CRASH("Not a variable modifier")
+		final = LAZYACCESS(actionspeed_modification, initial(type_id_datum.id) || "[type_id_datum]")
+		if(!final)
+			final = new type_id_datum
+			inject = TRUE
+			modified = TRUE
+	else
+		if(!initial(type_id_datum.variable))
+			CRASH("Not a variable modifier")
+		final = type_id_datum
+		if(!LAZYACCESS(actionspeed_modification, final.id))
+			inject = TRUE
+			modified = TRUE
+	if(!isnull(multiplicative_slowdown))
+		final.multiplicative_slowdown = multiplicative_slowdown
+		modified = TRUE
+	if(inject)
+		add_actionspeed_modifier(final, FALSE)
+	if(update && modified)
+		update_actionspeed(TRUE)
+	return final
+
+///Is there a actionspeed modifier for this mob
+/mob/proc/has_actionspeed_modifier(datum/actionspeed_modifier/datum_type_id)
+	var/key
+	if(ispath(datum_type_id))
+		key = initial(datum_type_id.id) || "[datum_type_id]"
+	else if(istext(datum_type_id))
+		key = datum_type_id
+	else
+		key = datum_type_id.id
+	return LAZYACCESS(actionspeed_modification, key)
+
+/// Go through the list of actionspeed modifiers and calculate a final actionspeed. ANY ADD/REMOVE DONE IN UPDATE_actionspeed MUST HAVE THE UPDATE ARGUMENT SET AS FALSE!
+/mob/proc/update_actionspeed()
+	. = 0
+	var/list/conflict_tracker = list()
+	for(var/key in get_actionspeed_modifiers())
+		var/datum/actionspeed_modifier/M = actionspeed_modification[key]
+		var/conflict = M.conflicts_with
+		var/amt = M.multiplicative_slowdown
+		if(conflict)
+			// Conflicting modifiers prioritize the larger slowdown or the larger speedup
+			// We purposefuly don't handle mixing speedups and slowdowns on the same id
+			if(abs(conflict_tracker[conflict]) < abs(amt))
+				conflict_tracker[conflict] = amt
+			else
+				continue
+		. += amt
+	cached_multiplicative_actions_slowdown = .
+
+///Adds a default action speed
+/mob/proc/initialize_actionspeed()
+	add_or_update_variable_actionspeed_modifier(/datum/actionspeed_modifier/base, multiplicative_slowdown = 1)
+
+/// Get the action speed modifiers list of the mob
+/mob/proc/get_actionspeed_modifiers()
+	. = LAZYCOPY(actionspeed_modification)
+	for(var/id in actionspeed_mod_immunities)
+		. -= id
+
+/// Checks if a action speed modifier is valid and not missing any data
+/proc/actionspeed_data_null_check(datum/actionspeed_modifier/M)		//Determines if a data list is not meaningful and should be discarded.
+	. = TRUE
+	if(M.multiplicative_slowdown)
+		. = FALSE
```

Comment:
Why not `return M.multiplicative_slowdown`?

Reply:
i like to pretend the language is statically typed so i sleep better at night

Reply:
`return M.multiplicative_slowdown > 0`?

Reply:
yeah ill add it :)